pkg/dns/serviceresolver.go

```diff
@@ -196,6 +201,16 @@ func (b *ServiceResolver) Records(dnsName string, exact bool) ([]msg.Service, er
 			return nil, errNoSuchName
 		}
 
+		hostnameMappings := noHostnameMappings
+		if savedHostnames := endpoints.Annotations[kendpoints.PodHostnamesAnnotation]; len(savedHostnames) > 0 {
+			mapped := make(map[string]kendpoints.HostRecord)
+			if err = json.Unmarshal([]byte(savedHostnames), &mapped); err == nil {
+				hostnameMappings = mapped
+			}
+		}
+
+		matchHostname := len(segments) > 3 && !hasAllPrefixedSegments(segments[3:4], "_")
```

Comment:
Also: nit: I think this would be clearer as `!strings.HasPrefix(segments[3], "_")` (the use of the single-element slice here seems ad bit odd to me...)